0.12.0-alpha4 crash in plan diff printer

<!--
Hi there,

Thank you for opening an issue. Please note that we try to keep the Terraform issue tracker reserved for bug reports and feature requests. For general usage questions, please see: https://www.terraform.io/community.html.

If your issue relates to a specific Terraform provider, please open it in the provider's own repository. The index of providers is at https://github.com/terraform-providers .
-->

### Terraform Version
<!---
Run `terraform -v` to show the version, and paste the result between the ``` marks below.

If you are not running the latest version of Terraform, please try upgrading because your issue may have already been fixed.
-->

```
Terraform v0.12.0-alpha4 (2c36829d3265661d8edbd5014de8090ea7e2a076)
+ provider.aws v1.40.0-6-gb23683732-dev
```

### Terraform Configuration Files
<!--
Paste the relevant parts of your Terraform configuration between the ``` marks below.

For large Terraform configs, please use a service like Dropbox and share a link to the ZIP file. For security, you can also encrypt the files using our GPG public key.
-->

https://github.com/querry43/terraform

### Debug Output
<!--
Full debug output can be obtained by running Terraform with the environment variable `TF_LOG=trace`. Please create a GitHub Gist containing the debug output. Please do _not_ paste the debug output in the issue, since debug output is long.

Debug output may contain sensitive information. Please review it before posting publicly, and if you are concerned feel free to encrypt the files using the HashiCorp security public key.
-->

### Crash Output
<!--
If the console output indicates that Terraform crashed, please share a link to a GitHub Gist containing the output of the `crash.log` file.
-->

https://gist.github.com/querry43/cc0b63e82ea015b08aff4864329fe04f

### Expected Behavior
<!--
What should have happened?
-->

It should not crash

### Actual Behavior
<!--
What actually happened?
-->

It crashed

### Steps to Reproduce
<!--
Please list the full steps required to reproduce the issue, for example:
1. `terraform init`
2. `terraform apply`
-->

1. `terraform init`
2. `terraform apply`

### Additional Context
<!--
Are there anything atypical about your situation that we should know? For example: is Terraform running in a wrapper script or in a CI system? Are you passing any unusual command line options or environment variables to opt-in to non-default behavior?
-->

### References
<!--
Are there any other GitHub issues (open or closed) or Pull Requests that should be linked here? For example:

- #6017

-->

Thanks for reporting this, @querry43.

```
panic: interface conversion: interface {} is *cty.unknownType, not bool

goroutine 40 [running]:
github.com/hashicorp/terraform/vendor/github.com/zclconf/go-cty/cty.Value.True(0x38cc820, 0xc0000ae501, 0x2d3a1c0, 0x63ba740, 0x38cc920)
	/opt/gopath/src/github.com/hashicorp/terraform/vendor/github.com/zclconf/go-cty/cty/value_ops.go:297 +0x115
github.com/hashicorp/terraform/command/format.(*blockBodyDiffPrinter).writeNestedBlockDiffs(0xc00133f780, 0xc000215090, 0x10, 0xc0007f72f0, 0x38cc960, 0xc000b56780, 0x343f800, 0xc00109c2a0, 0x38cc960, 0xc000b563e0, ...)
	/opt/gopath/src/github.com/hashicorp/terraform/command/format/diff.go:371 +0x1667
github.com/hashicorp/terraform/command/format.(*blockBodyDiffPrinter).writeBlockBodyDiff(0xc00133f780, 0xc0007ea370, 0x38cc920, 0xc001132c78, 0x2cfa3a0, 0xc001f8c6c0, 0x38cc920, 0xc001132c88, 0x2cfa3a0, 0xc001f8d830, ...)
	/opt/gopath/src/github.com/hashicorp/terraform/command/format/diff.go:195 +0xa26
github.com/hashicorp/terraform/command/format.ResourceChange(0xc000b92800, 0xc0007ea370, 0xc00031cde0, 0xc, 0xc00051a6f8)
	/opt/gopath/src/github.com/hashicorp/terraform/command/format/diff.go:125 +0x52b
github.com/hashicorp/terraform/backend/local.(*Local).renderPlan(0xc000387860, 0xc00032b650, 0xc00023c0c0)
	/opt/gopath/src/github.com/hashicorp/terraform/backend/local/backend_plan.go:254 +0x5ba
github.com/hashicorp/terraform/backend/local.(*Local).opApply(0xc000387860, 0x38cae60, 0xc00030e040, 0x38cae60, 0xc00030e080, 0xc000366380, 0xc00030e000)
	/opt/gopath/src/github.com/hashicorp/terraform/backend/local/backend_apply.go:111 +0x118b
github.com/hashicorp/terraform/backend/local.(*Local).opApply-fm(0x38cae60, 0xc00030e040, 0x38cae60, 0xc00030e080, 0xc000366380, 0xc00030e000)
	/opt/gopath/src/github.com/hashicorp/terraform/backend/local/backend.go:306 +0x66
github.com/hashicorp/terraform/backend/local.(*Local).Operation.func1(0xc000240520, 0xc000240530, 0xc000240540, 0xc000366380, 0xc000387860, 0xc00030e000, 0xc000240510, 0x38cae60, 0xc00030e040, 0x38cae60, ...)
	/opt/gopath/src/github.com/hashicorp/terraform/backend/local/backend.go:357 +0x135
created by github.com/hashicorp/terraform/backend/local.(*Local).Operation
	/opt/gopath/src/github.com/hashicorp/terraform/backend/local/backend.go:341 +0x362
```

Based on the backtrace here, it seems like this is an issue in the diff printer when Terraform is trying render the changeset for approval. Specifically, it seems like the renderer is not correctly handling an unknown boolean value somewhere. We'll take a look at this before final release and get it fixed.

--- a/command/format/diff.go
+++ b/command/format/diff.go
@@ -377,6 +377,9 @@ func (p *blockBodyDiffPrinter) writeNestedBlockDiffs(name string, blockS *config
 			var action plans.Action
 			var oldValue, newValue cty.Value
 			switch {
+			case !val.IsKnown():
+				action = plans.Update
+				newValue = val
 			case !old.HasElement(val).True():
 				action = plans.Create
 				oldValue = cty.NullVal(val.Type())
@@ -697,7 +700,7 @@ func (p *blockBodyDiffPrinter) writeValueDiff(old, new cty.Value, indent int, pa
 			for it := new.ElementIterator(); it.Next(); {
 				_, val := it.Element()
 				allVals = append(allVals, val)
-				if old.HasElement(val).False() {
+				if val.IsKnown() && old.HasElement(val).False() {
 					addedVals = append(addedVals, val)
 				}
 			}
@@ -726,6 +729,8 @@ func (p *blockBodyDiffPrinter) writeValueDiff(old, new cty.Value, indent int, pa
 
 				var action plans.Action
 				switch {
+				case !val.IsKnown():
+					action = plans.Update
 				case added.HasElement(val).True():
 					action = plans.Create
 				case removed.HasElement(val).True():
@@ -1062,7 +1067,7 @@ func ctyObjectSimilarity(old, new cty.Value) float64 {
 }
 
 func ctyEqualWithUnknown(old, new cty.Value) bool {
-	if !old.IsKnown() || !new.IsKnown() {
+	if !old.IsWhollyKnown() || !new.IsWhollyKnown() {
 		return false
 	}
 	return old.Equals(new).True()

--- a/command/format/diff_test.go
+++ b/command/format/diff_test.go
@@ -722,6 +722,92 @@ func TestResourceChange_primitiveList(t *testing.T) {
     }
 `,
 		},
+		"update to unknown element": {
+			Action: plans.Update,
+			Mode:   addrs.ManagedResourceMode,
+			Before: cty.ObjectVal(map[string]cty.Value{
+				"id":  cty.StringVal("i-02ae66f368e8518a9"),
+				"ami": cty.StringVal("ami-STATIC"),
+				"list_field": cty.ListVal([]cty.Value{
+					cty.StringVal("aaaa"),
+					cty.StringVal("bbbb"),
+					cty.StringVal("cccc"),
+				}),
+			}),
+			After: cty.ObjectVal(map[string]cty.Value{
+				"id":  cty.UnknownVal(cty.String),
+				"ami": cty.StringVal("ami-STATIC"),
+				"list_field": cty.ListVal([]cty.Value{
+					cty.StringVal("aaaa"),
+					cty.UnknownVal(cty.String),
+					cty.StringVal("cccc"),
+				}),
+			}),
+			Schema: &configschema.Block{
+				Attributes: map[string]*configschema.Attribute{
+					"id":         {Type: cty.String, Optional: true, Computed: true},
+					"ami":        {Type: cty.String, Optional: true},
+					"list_field": {Type: cty.List(cty.String), Optional: true},
+				},
+			},
+			RequiredReplace: cty.NewPathSet(),
+			ExpectedOutput: `  # test_instance.example will be updated in-place
+  ~ resource "test_instance" "example" {
+        ami        = "ami-STATIC"
+      ~ id         = "i-02ae66f368e8518a9" -> (known after apply)
+      ~ list_field = [
+            "aaaa",
+          - "bbbb",
+          + (known after apply),
+            "cccc",
+        ]
+    }
+`,
+		},
+		"update - two new unknown elements": {
+			Action: plans.Update,
+			Mode:   addrs.ManagedResourceMode,
+			Before: cty.ObjectVal(map[string]cty.Value{
+				"id":  cty.StringVal("i-02ae66f368e8518a9"),
+				"ami": cty.StringVal("ami-STATIC"),
+				"list_field": cty.ListVal([]cty.Value{
+					cty.StringVal("aaaa"),
+					cty.StringVal("bbbb"),
+					cty.StringVal("cccc"),
+				}),
+			}),
+			After: cty.ObjectVal(map[string]cty.Value{
+				"id":  cty.UnknownVal(cty.String),
+				"ami": cty.StringVal("ami-STATIC"),
+				"list_field": cty.ListVal([]cty.Value{
+					cty.StringVal("aaaa"),
+					cty.UnknownVal(cty.String),
+					cty.UnknownVal(cty.String),
+					cty.StringVal("cccc"),
+				}),
+			}),
+			Schema: &configschema.Block{
+				Attributes: map[string]*configschema.Attribute{
+					"id":         {Type: cty.String, Optional: true, Computed: true},
+					"ami":        {Type: cty.String, Optional: true},
+					"list_field": {Type: cty.List(cty.String), Optional: true},
+				},
+			},
+			RequiredReplace: cty.NewPathSet(),
+			ExpectedOutput: `  # test_instance.example will be updated in-place
+  ~ resource "test_instance" "example" {
+        ami        = "ami-STATIC"
+      ~ id         = "i-02ae66f368e8518a9" -> (known after apply)
+      ~ list_field = [
+            "aaaa",
+          - "bbbb",
+          + (known after apply),
+          + (known after apply),
+            "cccc",
+        ]
+    }
+`,
+		},
 	}
 	runTestCases(t, testCases)
 }

--- a/command/format/diff_test.go
+++ b/command/format/diff_test.go
@@ -1382,6 +1382,47 @@ func TestResourceChange_map(t *testing.T) {
     }
 `,
 		},
+		"update to unknown element": {
+			Action: plans.Update,
+			Mode:   addrs.ManagedResourceMode,
+			Before: cty.ObjectVal(map[string]cty.Value{
+				"id":  cty.StringVal("i-02ae66f368e8518a9"),
+				"ami": cty.StringVal("ami-STATIC"),
+				"map_field": cty.MapVal(map[string]cty.Value{
+					"a": cty.StringVal("aaaa"),
+					"b": cty.StringVal("bbbb"),
+					"c": cty.StringVal("cccc"),
+				}),
+			}),
+			After: cty.ObjectVal(map[string]cty.Value{
+				"id":  cty.UnknownVal(cty.String),
+				"ami": cty.StringVal("ami-STATIC"),
+				"map_field": cty.MapVal(map[string]cty.Value{
+					"a": cty.StringVal("aaaa"),
+					"b": cty.UnknownVal(cty.String),
+					"c": cty.StringVal("cccc"),
+				}),
+			}),
+			Schema: &configschema.Block{
+				Attributes: map[string]*configschema.Attribute{
+					"id":        {Type: cty.String, Optional: true, Computed: true},
+					"ami":       {Type: cty.String, Optional: true},
+					"map_field": {Type: cty.Map(cty.String), Optional: true},
+				},
+			},
+			RequiredReplace: cty.NewPathSet(),
+			ExpectedOutput: `  # test_instance.example will be updated in-place
+  ~ resource "test_instance" "example" {
+        ami       = "ami-STATIC"
+      ~ id        = "i-02ae66f368e8518a9" -> (known after apply)
+      ~ map_field = {
+            "a" = "aaaa"
+          ~ "b" = "bbbb" -> (known after apply)
+            "c" = "cccc"
+        }
+    }
+`,
+		},
 	}
 	runTestCases(t, testCases)
 }

--- a/command/format/diff_test.go
+++ b/command/format/diff_test.go
@@ -1004,6 +1004,80 @@ func TestResourceChange_primitiveSet(t *testing.T) {
     }
 `,
 		},
+		"in-place update to unknown": {
+			Action: plans.Update,
+			Mode:   addrs.ManagedResourceMode,
+			Before: cty.ObjectVal(map[string]cty.Value{
+				"id":  cty.StringVal("i-02ae66f368e8518a9"),
+				"ami": cty.StringVal("ami-STATIC"),
+				"set_field": cty.SetVal([]cty.Value{
+					cty.StringVal("aaaa"),
+					cty.StringVal("bbbb"),
+				}),
+			}),
+			After: cty.ObjectVal(map[string]cty.Value{
+				"id":        cty.UnknownVal(cty.String),
+				"ami":       cty.StringVal("ami-STATIC"),
+				"set_field": cty.UnknownVal(cty.Set(cty.String)),
+			}),
+			Schema: &configschema.Block{
+				Attributes: map[string]*configschema.Attribute{
+					"id":        {Type: cty.String, Optional: true, Computed: true},
+					"ami":       {Type: cty.String, Optional: true},
+					"set_field": {Type: cty.Set(cty.String), Optional: true},
+				},
+			},
+			RequiredReplace: cty.NewPathSet(),
+			ExpectedOutput: `  # test_instance.example will be updated in-place
+  ~ resource "test_instance" "example" {
+        ami       = "ami-STATIC"
+      ~ id        = "i-02ae66f368e8518a9" -> (known after apply)
+      ~ set_field = [
+          - "aaaa",
+          - "bbbb",
+        ] -> (known after apply)
+    }
+`,
+		},
+		"in-place update to unknown element": {
+			Action: plans.Update,
+			Mode:   addrs.ManagedResourceMode,
+			Before: cty.ObjectVal(map[string]cty.Value{
+				"id":  cty.StringVal("i-02ae66f368e8518a9"),
+				"ami": cty.StringVal("ami-STATIC"),
+				"set_field": cty.SetVal([]cty.Value{
+					cty.StringVal("aaaa"),
+					cty.StringVal("bbbb"),
+				}),
+			}),
+			After: cty.ObjectVal(map[string]cty.Value{
+				"id":  cty.UnknownVal(cty.String),
+				"ami": cty.StringVal("ami-STATIC"),
+				"set_field": cty.SetVal([]cty.Value{
+					cty.StringVal("aaaa"),
+					cty.UnknownVal(cty.String),
+				}),
+			}),
+			Schema: &configschema.Block{
+				Attributes: map[string]*configschema.Attribute{
+					"id":        {Type: cty.String, Optional: true, Computed: true},
+					"ami":       {Type: cty.String, Optional: true},
+					"set_field": {Type: cty.Set(cty.String), Optional: true},
+				},
+			},
+			RequiredReplace: cty.NewPathSet(),
+			ExpectedOutput: `  # test_instance.example will be updated in-place
+  ~ resource "test_instance" "example" {
+        ami       = "ami-STATIC"
+      ~ id        = "i-02ae66f368e8518a9" -> (known after apply)
+      ~ set_field = [
+            "aaaa",
+          - "bbbb",
+          ~ (known after apply),
+        ]
+    }
+`,
+		},
 	}
 	runTestCases(t, testCases)
 }

